File: src/mcp/relay.ts
import type { EventBus } from "../core/event-bus.js";

export interface RelayOptions {
	url: string;
	apiKey: string;
	fetchImpl?: typeof fetch;
}

/**
 * Subscribe to the MCP process's event bus and POST each change to the HTTP
 * server's internal relay endpoint. Fire-and-forget: a failed POST is logged
 * to stderr but never propagates into the tool call.
 *
 * Returns an unsubscribe function.
 */
export function startChangeRelay(
	bus: EventBus,
	opts: RelayOptions,
): () => void {
	const { url, apiKey, fetchImpl = fetch } = opts;

	return bus.subscribe((ev) => {
		fetchImpl(url, {
			method: "POST",
			headers: {
				authorization: `Bearer ${apiKey}`,
				"content-type": "application/json",
			},
			body: JSON.stringify({ type: ev.type, data: ev.data }),
		}).catch((err) => {
			console.error("[relay] failed to forward change:", err);
		});
	});
}

File: src/mcp/relay.rs
// ports: src/mcp/relay.ts
use crate::core::event_bus::EventBus;
use tokio::sync::broadcast;

pub struct RelayConfig {
    pub url: String,
    pub api_key: String,
}

pub fn start_relay(bus: &EventBus, config: RelayConfig) {
    let mut rx = bus.subscribe();
    let client = reqwest::Client::new();

    tokio::spawn(async move {
        loop {
            match rx.recv().await {
                Ok(event) => {
                    let body = serde_json::json!({
                        "type": event.event_type,
                        "data": event.data,
                    });
                    let res = client
                        .post(&config.url)
                        .header("authorization", format!("Bearer {}", config.api_key))
                        .header("content-type", "application/json")
                        .json(&body)
                        .send()
                        .await;
                    if let Err(e) = res {
                        eprintln!("[relay] failed to forward change: {e}");
                    }
                }
                Err(broadcast::error::RecvError::Lagged(n)) => {
                    eprintln!("[relay] lagged, skipped {n} events");
                }
                Err(broadcast::error::RecvError::Closed) => break,
            }
        }
    });
}
